### ISSUE TITLE
support short lived signing certificates (proof-of-concept)

### ISSUE THREAD

**Issue by jku** [secure-systems-lab/securesystemslib#399]
These are more like notes to self but _maybe_ useful for someone else who wants to experiment.

I've had a look at how python-tuf could use short lived signing cerfificates (as in Sigstore Fulcio) and came to this conclusion:
* the signing part  is a bit challenging
  * both OIDC authentication and the Fulcio certificate request can be a bit of work -- and obviously require connecting to services so require a design that's not usual for securesystemslib
  * https://github.com/trailofbits/sigstore-python seems to maybe be looking at this already: let's not invent multiple wheels
* however, the verification side is something we could deal with in securesystemslib already:
  * This would be the client half of the reference implementation of https://github.com/theupdateframework/taps/pull/141
  * This definitely requires dependencies, at least for an easy implementation: cryptography alone is not enough (it does not implement certificate chain validation) but cryptography+pyopenssl would work
  * We could test this by constructing example metadata manually (probably a bit painfully) using sigstore cosign to sign some metadata blobs
  
## Basic Verification process

Assumptions:
 * root certificate and signer identity (email) are embedded in the input key structure (a new keytype that includes these fields)
 * signing certificate is embedded in the signature structure (a new optional field in signature structure)

With these assumptions verify_signature() does not need API changes.

verify_signature() process:
1. Assert signing certificate is signed by root cert
2. Assert that signing certificate is a code signing certificate for the signer identity
3. Assert that we support the key type of the public key in the signing certificate
4. Assert that give input data is correctly signed by the public key in the certificate

## rough example

I've got some pyopenssl + cryptography code that I used to do a very rough pass of the first steps above (this is not securesystemslib related, I just tried to ensure that these libs can be made to do the job):

    # signer_id would in reality come from key structure
    signer_id = "email@example.com"

    # Load the signing cert and root cert (these would in reality be in signature and key structure respectively)
    from cryptography import x509
    pem_bytes = open("signing_cert.pem", "rb").read()
    cert = x509.load_pem_x509_certificate(pem_bytes)
    pem_bytes = open("fulcio.crt.pem", "rb").read()
    root = cryptography.x509.load_pem_x509_certificate(pem_bytes)

    # 1. Code to check that certificate is signed by root certificate

    # cryptography does not support this: need pyopenssl
    from OpenSSL import crypto
    sign_date = cert.not_valid_before
    openssl_cert = crypto.X509.from_cryptography(cert)
    openssl_root = crypto.X509.from_cryptography(root)
    
    # verify certificate validity at signing time
    store = crypto.X509Store()
    store.add_cert(openssl_root)
    store.set_time(sign_date)
    store_ctx = crypto.X509StoreContext(store, openssl_cert)
    store_ctx.verify_certificate()

    # 2.Assert that signing cert is for the signer identity

    # check usage is "digital signature"
    usage_ext = cert.extensions.get_extension_for_class(x509.KeyUsage)
    assert usage_ext.value.digital_signature

    # check that extended usage contains "code signing"
    extended_usage_ext = cert.extensions.get_extension_for_class(x509.ExtendedKeyUsage)
    assert x509.oid.ExtendedKeyUsageOID.CODE_SIGNING in extended_usage_ext.value

    # check that SubjectAlternativeName contains signer identity
    san_ext = cert.extensions.get_extension_for_class(x509.SubjectAlternativeName)
    assert signer_id in san_ext.value.get_values_for_type(x509.RFC822Name)

    # Get the public key:
    from cryptography.hazmat.primitives.serialization.Encoding import PEM
    from cryptography.hazmat.primitives.serialization.PublicFormat import SubjectPublicKeyInfo

    pubkey = cert.public_key()
    pem = pubkey.public_bytes(PEM, SubjectPublicKeyInfo)
    

    # ... the rest should be as usual: use pem as a public key, verify that signature is valid.

**Comment by trishankatdatadog** [comment]
I like it. So far it's been theoretical and unclear enough for a complete understanding.

**Comment by jku** [comment]
> We could test this by constructing example metadata manually (probably a bit painfully) using sigstore cosign to sign some metadata blobs

* Generate delegating metadata and delegated metadata. Use a a Fulcio-compatible key (ecdsa-sha2-nistp384 **should** work) for delegated metadata. Make sure delegated metadata is signed
* Use sigstore cosign to create the signing certificate, feeding in the key from step 1 (private key can now be deleted)
* Modify delegating metadata
  * delegated key should be a "fulcio" key with just a root cert and issuer
  * sign delegating metadata with normal means
* Modify delegated Metadata
  * keep the signature, but add signing certificate
  * no signing required: the content did not change and is already signed by the correct key

**Comment by jku** [comment]
> Use sigstore cosign to create the signing certificate, feeding in the key from step 1 

this does not actually work as cosign either uses given keys or uses OIDC+fulcio: it's not possible to use OIDC-fulcio and "override" the ephemeral key with my own.

So the remaining option for creating metadata (before we have proper signing support) is to use cosign with OIDC+fulcio and to feed it canonical JSON and hope that it all happens in a compatible way...

**Comment by jku** [comment]
https://github.com/sigstore/sigstore-python/pull/56 sigstore-python just added OIDC support via browser ... should be able to leverage that to create a rough signing POC

**Comment by lukpueh** [comment]
Times flies... sigstore-python has a stable API these days, and securesystemslib has a new signer architecture. I think we can close this courtesy of #522, #535, etc. Need to check, if any notes/thoughts from here should be retained, though.

**Comment by jku** [comment]
No, I think all of the details in the code are handled by sigstore-python :v:
